File: src/python/pythonsearchengine/auto_complete.py
# we will Trie data structure to make the auto_complete feature

class TrieNode():
    def __init__(self):
        self.children = {}
        self.is_word = False

class Trie():
    def __init__(self, words_dict) -> None:
        self.root = TrieNode()
        self.words_dict = words_dict

    def createTrie(self):

        for word in self.words_dict:
            self.insert(word)

    def insert(self, word):

        node = self.root

        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()

            node = node.children[char]

        node.is_word = True

class AutoComplete(Trie):
    def __init__(self, words_dict) -> None:
        super().__init__(words_dict)

    def autocomplete(self, node, word):

        results = []
        if node.is_word:
            results.append(word)

        for char, childnode in node.children.items():
            results.extend(self.autocomplete(childnode, word + char))

        return results

    def calc_score(self, word):
        return self.words_dict[word]

    def get_suggestion(self, word):

        ans = []

        node = self.root

        for char in word:

            if char not in node.children:
                return (ans, 0)

            node = node.children[char]


        if not node.children:
            return (ans, -1)

        # we need at most 10-highest-score suggestions
        ans = sorted(self.autocomplete(node, word), key = self.calc_score)[:10]

        return (ans, 1)

File: src/main/java/javasearchengine/core/autocomplete/AutoComplete.java
package javasearchengine.core.autocomplete;

import java.util.*;

public class AutoComplete extends Trie {

    public AutoComplete(Map<String, Integer> wordDict) {
        super(wordDict);

    }

    public List<String> autoComplete(TrieNode node, String word) {
        List<String> results = new ArrayList<>();
        if (node.getCheckIsWord() == true) {
            results.add(word);
        }
        for (Map.Entry<Character, TrieNode> entry : node.getChildren().entrySet()) {
            for (String newWord : this.autoComplete(entry.getValue(), word + entry.getKey())) {

                results.add(newWord);

            }
        }
        return results;
    }

    public int calcScore(String word) {
        return super.getWordDict().get(word);
    }

    public Map<List<String>, Integer> getSuggestion(String word) {

        Map<List<String>, Integer> res = new HashMap<>();

        List<String> ans = new ArrayList<>();

        TrieNode node = super.getRoot();

        for (char c : word.toCharArray()) {

            if (!node.getChildren().containsKey(c)) {
                res.put(ans, 0);
                return res;
            }

            node = node.getChildren().get(c);
        }

        if (node.getChildren() == null) {
            res.put(ans, -1);
            return res;
        }

        List<String> listSuggest = this.autoComplete(node, word);

        // we need at most 10-highest-score suggestions
        Collections.sort(listSuggest, new Comparator<String>() {

            @Override
            public int compare(String word1, String word2) {
                // TODO Auto-generated method stub
                return -calcScore(word1) + calcScore(word2);
            }

        });

        int cnt = 0;
        for (String w : listSuggest) {
            ans.add(w);
            cnt++;
            if (cnt == 10) {
                break;
            }
        }
        res.put(ans, 1);
        return res;

    }

}